src/components/data_submission/ds_common.js

```diff
@@ -0,0 +1,341 @@
+import { h, div, label, input, span, button } from 'react-hyperscript-helpers';
+import { cloneDeep, get, set } from 'lodash/fp';
+import { SearchSelectOrText } from '../SearchSelectOrText';
+import { Theme } from '../../libs/theme';
+import './ds_common.css';
+
+export const styles = {
+  inputStyle: {
+    padding: '25px 15px',
+    width: '100%'
+  },
+  flexRow: {
+    display: 'flex',
+    flexDirection: 'row',
+    justifyContent: 'space-between',
+    width: '100%'
+  }
+};
+
+//---------------------------------------------
+// Form Behavior
+//---------------------------------------------
+const validateFormInput = (config, value) => {
+  const {
+    id, errors, setErrors,
+    required, isValid
+  } = config;
+
+  const valueExists = value !== undefined && value !== null && value !== '';
+  const requiredValid = !required || (required && valueExists);
+  const customValid = !isValid || isValid(value);
+  const isValidInput = requiredValid && customValid;
+  const errorsClone = cloneDeep(errors);
+
+  if (isValidInput) {
+    delete errorsClone[id];
+    setErrors(errorsClone);
+  } else {
+    errorsClone[id] = true;
+    setErrors(errorsClone);
+  }
+
+  return isValidInput;
+};
+
+const normalizeValue = (value) => {
+  if (typeof value === 'string') {
+    return value.trim();
+  } else if (Array.isArray(value)) {
+    return value
+      .map(x => normalizeValue(x))
+      .filter(x => x); // filter out null strings
+  }
+  return value;
+};
+
+const onFormInputChange = (config, value) => {
+  const { id, onChange, formInfo, setFormInfo, path } = config;
+  const normalizedValue = normalizeValue(value);
+  const isValidInput = validateFormInput(config, normalizedValue);
+
+  if (isValidInput) {
+    onChange({key: path || id, value: normalizedValue});
+    const formInfoClone = cloneDeep(formInfo);
+    const updatedFormInfo = set((path || id), normalizedValue, formInfoClone);
+    setFormInfo(updatedFormInfo);
+  }
+};
+
+//---------------------------------------------
+// Form Controls
+//---------------------------------------------
+const formInput = (config) => {
+  switch (config.type) {
+    case 'select': return formInputSelect(config);
```

Comment:
maybe we could make these strings constants and export them so that you don't ever have to use a raw string to specify things